tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AdoNet/DbScopeFactory.cs

```diff
@@ -68,14 +70,35 @@ private static Scope CreateDbCommandScope(Tracer tracer, IDbCommand command, Int
                 if (tracer.Settings.DbmPropagationMode != DbmPropagationLevel.Disabled
                     && command.CommandType != CommandType.StoredProcedure)
                 {
-                    var propagatedCommand = DatabaseMonitoringPropagator.PropagateSpanData(tracer.Settings.DbmPropagationMode, tracer.DefaultServiceName, scope.Span, integrationId, out var traceParentInjected);
-
-                    if (propagatedCommand != string.Empty)
+                    var alreadyInjected = commandText.StartsWith(DatabaseMonitoringPropagator.DbmPrefix);
+                    if (alreadyInjected)
+                    {
+                        // The command text is already injected, so they're probably caching the SqlCommand
+                        // that's not a problem if they're using 'service' mode, but it _is_ a problem for 'full' mode
+                        // There's not a lot we can do about it (we don't want to start parsing commandText), so just
+                        // report it for now
+                        if (!Volatile.Read(ref _dbCommandCachingLogged)
+                            && tracer.Settings.DbmPropagationMode != DbmPropagationLevel.Service)
+                        {
+                            _dbCommandCachingLogged = true;
```

Comment:
I struggle with how to properly use `volatile` and `Volatile.Read`/`Volatile.Write`, but should this be a `Volatile.Write(ref _dbCommandCachingLogged, true)`?
Also would using `Interlocked.CompareExchange` work instead of doing the `Volatile.Read` or is the memory barrier stuff of the `Volatile.Read` the goal here?

Reply:
Haha, you and me both. As I understand it, writes always have a memory barrier, so that's everything we need. Really need @kevingosse to swoop in and put as straight on this one 😄